README.md

```diff
@@ -1,12 +1,264 @@
+- `1.0.0`: file tree with metadata. Nodes in the file tree are structured as 2 layers where one layer contains "header" information (metadata, cache, etc), and the second layer contains data or links. 
+**This is the default version, use this unless you have a good reason not to**
+- `0.0.0`: bare file tree. The public tree is [ipfs dag-pg](https://github.com/ipld/js-ipld-dag-pb) nodes. The private tree is encrypted links with no associated metadata. These shuld really only be used for vanity links to be rendered by a gateway.
+
+## API
+
+### Config
+Each instantiation method takes an optional config. Below is the default config and descriptions of each value. 
+```ts
+const defaultConfig = {
+  keyName: 'filesystem-root', // the name of the key for the filesystem root as stored in IndexedDB
+  version: '1.0.0' // the version of the filesystem as discussed above
+}
+```
+--- 
+
+### Instantiation
+
+**empty**
+
+Creates a file system with an empty public tree & an empty private tree at the root
+
+Params: 
+- cfg: `FileSystemConfig` _optional_
+
+Example:
+```ts
+import FileSystem from 'fission-sdk/ffs'
+const wnfs = await FileSystem.empty()
+```
+
+---
+
+**fromCID**
+
+Loads an existing file system from a CID
+
+Params: 
+- cid: `CID` (`string`) **required**
+- cfg: `FileSystemConfig` _optional_
+
+Example:
+```ts
+import FileSystem from 'fission-sdk/ffs'
+const cid = "QmWKst5WVNTPfMsSFCQEJYBJLEsUZfghqrKXJBVU4EuA76"
+const wnfs = await FileSystem.empty(cid)
+```
+
+---
+
+**forUser**
+
+Loads an existing file system from a username
+
+Params: 
+- username: `string` **required**
+- cfg: `FileSystemConfig` _optional_
+
+Example:
+```ts
+import FileSystem from 'fission-sdk/ffs'
+const wnfs = await FileSystem.foruser("boris")
+```
+
+---
+
+### Methods
+Methods for interacting with the filesystem all use **absolute** paths. We're planning on adding a [stateful session](https://github.com/fission-suite/ts-sdk/issues/24) but for now, filesystem state will need to be tracked in your application.
+
+**ls**
+
+Returns a list of links at a given directory path
+
+Params: 
+- path: `string` **required**
+
+Example:
+```ts
+// public
+const links = await wnfs.ls("public/some/directory/path")
+// private
+const links = await wnfs.ls("private/some/directory/path")
+```
+
+---
+
+**mkdir**
+
+Creates a directory at the given path
+
+Params: 
+- path: `string` **required**
+
+Example:
+```ts
+const updatedCID = await wnfs.mkdir("public/some/directory/path")
+// creates a directory called "path" at "public/some/directory"
+```
+
+---
+
+**add**
+
+Adds some file content at a given path
+
+Params: 
+- path: `string` **required**
+- content: `FileContent` (`object | string | Blob`) **required**
+
+Example:
+```ts
+const content = "hello world"
+const updatedCID = await wnfs.add("public/some/path/to/a/file", content)
+// creates a file called "file" at "public/some/path/to/a"
+```
+
+---
+
+**cat**
+
+Retrieves some file content at a given path
```

Comment:
Might be good to mention the return type here. Does this return a blob or an array buffer?

Reply:
Well data in the private tree is cbor encoded before being uploaded, so from the private tree `cat` can return deserialized info like an object, array, string etc. From the public tree, everything comes out as a `Buffer` (or `Uint8Array` in the browse). So it returns type `FileContent (object | string | Blob | Buffer)`.

I'm really not quite sure the best way to handle this... Standardize it around always return a `Buffer` or return a deserialized object when possible (so the vague `FileContent` type)?

Regardless, I took your advice on adding return types :+1:

Reply:
Hmm, I'd say standardize on one format. Not sure if that format would be a Buffer, Blob or Async Iterable 🤷